File: client/android/app/src/main/java/com/clipbase/ApiClient.kt
package com.clipbase

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.net.URLEncoder
import java.util.UUID

object ApiClient {
    /// 服务器地址：改成 Windows 电脑的局域网 IP（真机 + 电脑同一 WiFi）
    private const val BASE_URL = "http://192.168.3.36:3000"

    private val client = OkHttpClient()
    private val jsonType = "application/json; charset=utf-8".toMediaType()

    fun userId(context: Context): String {
        val prefs = context.getSharedPreferences("clipbase", Context.MODE_PRIVATE)
        var id = prefs.getString("user_id", null)
        if (id == null) {
            id = UUID.randomUUID().toString()
            prefs.edit().putString("user_id", id).apply()
        }
        return id
    }

    private suspend fun call(method: String, path: String, body: JSONObject? = null): JSONObject =
        withContext(Dispatchers.IO) {
            val rb = body?.toString()?.toRequestBody(jsonType)
            val builder = Request.Builder().url("$BASE_URL$path")
            when (method) {
                "POST" -> builder.post(rb ?: "{}".toRequestBody(jsonType))
                "PATCH" -> builder.patch(rb ?: "{}".toRequestBody(jsonType))
            }
            client.newCall(builder.build()).execute().use { resp ->
                if (!resp.isSuccessful) throw IOException("HTTP ${resp.code}")
                val text = resp.body?.string() ?: "{}"
                if (text.isBlank()) JSONObject() else JSONObject(text)
            }
        }

    // 条目

    suspend fun fetchItems(context: Context, category: String? = null, sceneId: String? = null, q: String? = null): List<Item> {
        val params = mutableListOf("user_id=${userId(context)}")
        if (!category.isNullOrEmpty()) params.add("category=${URLEncoder.encode(category, "UTF-8")}")
        if (!sceneId.isNullOrEmpty()) params.add("scene_id=${URLEncoder.encode(sceneId, "UTF-8")}")
        if (!q.isNullOrEmpty()) params.add("q=${URLEncoder.encode(q, "UTF-8")}")
        val json = call("GET", "/api/items?${params.joinToString("&")}")
        val arr = json.getJSONArray("items")
        return (0 until arr.length()).map { Item.fromJson(arr.getJSONObject(it)) }
    }

    suspend fun fetchItem(id: String): Item = Item.fromJson(call("GET", "/api/items/$id"))

    suspend fun ingest(context: Context, text: String?, url: String?) {
        val body = JSONObject().apply {
            put("user_id", userId(context))
            if (!text.isNullOrEmpty()) put("text", text)
            if (!url.isNullOrEmpty()) put("url", url)
        }
        call("POST", "/api/ingest", body)
    }

    suspend fun updateItem(id: String, category: String, tags: List<String>, intent: String, digest: String) {
        val body = JSONObject().apply {
            put("category", category)
            put("tags", JSONArray(tags))
            put("intent", intent)
            put("digest_state", digest)
        }
        call("PATCH", "/api/items/$id", body)
    }

    suspend fun retry(id: String) = call("POST", "/api/items/$id/retry")

    suspend fun assignItemToScene(id: String, sceneId: String?) {
        call("POST", "/api/items/$id/scene", JSONObject().put("scene_id", sceneId ?: ""))
    }

    suspend fun similarAction(id: String, action: String) {
        call("POST", "/api/items/$id/similar-action", JSONObject().put("action", action))
    }

    // 场景

    suspend fun fetchScenes(context: Context): List<Scene> {
        val json = call("GET", "/api/scenes?user_id=${userId(context)}")
        val arr = json.getJSONArray("scenes")
        return (0 until arr.length()).map { Scene.fromJson(arr.getJSONObject(it)) }
    }

    suspend fun createScene(context: Context, name: String) {
        call("POST", "/api/scenes", JSONObject().put("user_id", userId(context)).put("name", name))
    }

    // 创造

    suspend fun compose(context: Context, type: String): String {
        val json = call("POST", "/api/compose", JSONObject().put("user_id", userId(context)).put("type", type))
        return json.optString("content")
    }

    suspend fun weekly(context: Context): String {
        val json = call("GET", "/api/weekly?user_id=${userId(context)}")
        return json.optString("content")
    }

    suspend fun fetchDashboard(context: Context): Dashboard {
        val json = call("GET", "/api/dashboard?user_id=${userId(context)}")
        return Dashboard.fromJson(json)
    }
}

File: client/ios/Clipbase/Services/APIClient.swift
import Foundation

/// 后端 API 客户端
struct APIClient {
    static let shared = APIClient()

    /// 服务器地址：
    /// - 后端跑在同一台 Mac 上：http://127.0.0.1:3000
    /// - 后端跑在别的机器：改成那台机器局域网 IP
    static let baseURL = URL(string: "http://127.0.0.1:3000")!

    static let appGroup = "group.com.clipbase"

    static var userId: String {
        let defaults = UserDefaults(suiteName: appGroup)!
        if let id = defaults.string(forKey: "user_id") { return id }
        let newId = UUID().uuidString
        defaults.set(newId, forKey: "user_id")
        return newId
    }

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()

    private func request(_ path: String, method: String = "GET", body: [String: Any]? = nil) async throws -> Data {
        var req = URLRequest(url: APIClient.baseURL.appendingPathComponent(path))
        req.httpMethod = method
        if let body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, _) = try await URLSession.shared.data(for: req)
        return data
    }

    // MARK: 条目

    func fetchItems(category: String? = nil, sceneId: String? = nil, q: String? = nil) async throws -> [Item] {
        var comps = URLComponents(url: APIClient.baseURL.appendingPathComponent("api/items"), resolvingAgainstBaseURL: false)!
        var query = [URLQueryItem(name: "user_id", value: APIClient.userId)]
        if let category, !category.isEmpty { query.append(URLQueryItem(name: "category", value: category)) }
        if let sceneId, !sceneId.isEmpty { query.append(URLQueryItem(name: "scene_id", value: sceneId)) }
        if let q, !q.isEmpty { query.append(URLQueryItem(name: "q", value: q)) }
        comps.queryItems = query
        let (data, _) = try await URLSession.shared.data(from: comps.url!)
        return try decoder.decode(ItemListResponse.self, from: data).items
    }

    func fetchItem(id: String) async throws -> Item {
        try decoder.decode(Item.self, from: await request("api/items/\(id)"))
    }

    func ingest(text: String? = nil, url: String? = nil) async throws {
        var body: [String: Any] = ["user_id": APIClient.userId]
        if let text { body["text"] = text }
        if let url { body["url"] = url }
        _ = try await request("api/ingest", method: "POST", body: body)
    }

    func updateItem(id: String, category: String?, tags: [String]?, intent: String? = nil, digestState: String? = nil) async throws -> Item {
        var body: [String: Any] = [:]
        if let category { body["category"] = category }
        if let tags { body["tags"] = tags }
        if let intent { body["intent"] = intent }
        if let digestState { body["digest_state"] = digestState }
        return try decoder.decode(Item.self, from: await request("api/items/\(id)", method: "PATCH", body: body))
    }

    func retry(id: String) async throws {
        _ = try await request("api/items/\(id)/retry", method: "POST")
    }

    func assignItem(id: String, sceneId: String?) async throws -> Item {
        try decoder.decode(Item.self, from: await request("api/items/\(id)/scene", method: "POST", body: ["scene_id": sceneId ?? ""]))
    }

    func similarAction(id: String, action: String) async throws {
        _ = try await request("api/items/\(id)/similar-action", method: "POST", body: ["action": action])
    }

    // MARK: 场景

    func fetchScenes() async throws -> [Scene] {
        let data = try await request("api/scenes?user_id=\(APIClient.userId)")
        struct Resp: Codable { let scenes: [Scene] }
        return try decoder.decode(Resp.self, from: data).scenes
    }

    func createScene(name: String, emoji: String? = nil) async throws {
        var body: [String: Any] = ["user_id": APIClient.userId, "name": name]
        if let emoji { body["emoji"] = emoji }
        _ = try await request("api/scenes", method: "POST", body: body)
    }

    // MARK: 搜索 + 创造

    func search(q: String) async throws -> [Item] {
        struct R: Codable { let results: [Item] }
        let data = try await request("api/search?user_id=\(APIClient.userId)&q=\(q.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? q)")
        return try decoder.decode(R.self, from: data).results
    }

    func compose() async throws -> String {
        struct R: Codable { let title: String; let content: String }
        let data = try await request("api/compose", method: "POST", body: ["user_id": APIClient.userId])
        return try decoder.decode(R.self, from: data).content
    }

    func weekly() async throws -> String {
        struct R: Codable { let content: String }
        let data = try await request("api/weekly?user_id=\(APIClient.userId)")
        return try decoder.decode(R.self, from: data).content
    }
}
